modules/redhen_connection/src/ConnectionService.php

```diff
@@ -262,88 +246,85 @@ public function getAllConnectionEntityTypes() {
+      foreach ($connection_types as $type) {
+        if ($endpoint_fields = $type->getEndpointFields($entity_type)) {
+          $potential_endpoints[$type->id()]['entity1'] = $endpoint_fields;
+        }
 
-    // Overall OR group of connection_type/endpoint groupings.
-    $endpoints_group = $query->orConditionGroup();
+        if ($entity2_type) {
+          if ($endpoint2_fields = $type->getEndpointFields($entity2_type)) {
+            $potential_endpoints[$type->id()]['entity2'] = $endpoint2_fields;
+          }
+        }
+      }
 
-    // @todo Might instead be able to query against endpoint_1.entity.type, etc.
-    // Build endpoint groups.
-    foreach ($types as $type => $connection_type) {
-      /** @var \Drupal\redhen_connection\Entity\ConnectionTypeInterface $connection_type */
-      $endpoints = [];
-      $endpoints[$entity_type] = $connection_type->getEndpointFields($entity_type);
+      $database = \Drupal::database();
 
-      // Add condition for the connection_type.
-      $condition_group = $query->andConditionGroup()
-        ->condition('type', $type);
+      foreach ($potential_endpoints as $connection_type => $endpoint_group) {
 
-      // Working with 2 endpoints.
-      if ($entity_type2) {
-        $group = $query->andConditionGroup();
-        $endpoints[$entity_type2] = $connection_type->getEndpointFields($entity_type2);
+        $query = $database->select('redhen_connection', 'rc')
+          ->fields('rc', ['id'])
+          ->condition('type', $connection_type);
 
-        foreach ($endpoints as $endpoint_type => $endpoint_fields) {
-          for ($x = 0; $x < count($endpoint_fields); $x++) {
-            $group->condition($endpoint_fields[$x], $entities[$endpoint_type][$x]->id());
-          }
-          // Endpoints are of the same type so we need to add an additional
-          // condition for the reverse structure.
-          if ($x > 1) {
-            $group2 = $query->orConditionGroup();
-            for ($x = count($endpoint_fields) - 1; $x >= 0; $x--) {
-              $group2->condition($endpoint_fields[$x], $entities[$endpoint_type][$x]->id());
-            }
-            $group = $query->orConditionGroup()
-              ->condition($group)
-              ->condition($group2);
-          }
+        if ($active) {
+          $query->condition('status', $active);
         }
 
-        $condition_group->condition($group);
-      }
-      else {
-        // Single entity.
-        $condition_group = $query->orConditionGroup();
-        foreach ($endpoints as $endpoint_type => $endpoint_fields) {
-          for ($x = 0; $x < count($endpoint_fields); $x++) {
-            $condition_group->condition($endpoint_fields[$x], $entity->id());
+        // Parent condition group.
+        $entityAndGroup = $query->andConditionGroup();
+
+        // Entity 1 Group.
+        $entity1Group = $query->orConditionGroup();
+        $entity1Group->condition($endpoint_group['entity1'][0], $entity->id());
```

Comment:
The function definition for `getEndpointFields()` doesn't promise us that we'll have numerical indexes starting at 0, so it would be ideal not to rely on that. Would it be reasonable to loop on `$endpoint_group['entity1']` instead of doing these independent checks with the same call to `->condition` on 2 different lines?

Reply:
Good point! I updated the logic to not plan on that key.